### PR TITLE
Add ediff1d

### PR DESCRIPTION
**Pull request by jakirkham**
Implements NumPy's [ediff1d]( https://docs.scipy.org/doc/numpy/reference/generated/numpy.ediff1d.html ) for use with Dask Arrays. Includes tests against the original implementation with a variety of parameters. Also adds the function to the Dask Array API docs.

### CODE REVIEW

**Comment by jakirkham** [comment]
Looks like one of the workers crashed in this [build]( https://travis-ci.org/dask/dask/jobs/265217503 ). Not seeing how the changes here would have caused that. A restart would be appreciated.

File: dask/array/routines.py [hunk 1]
```diff
@@ -183,6 +183,23 @@ def diff(a, n=1, axis=-1):
...
+
+    aryf = ary.flatten()
+    r = aryf[1:] - aryf[:-1]
+
+    r = [r]
+    if to_begin is not None:
+        r = [asarray(to_begin).flatten()] + r
+    if to_end is not None:
+        r = r + [asarray(to_end).flatten()]
+    r = concatenate(r)
```
Comment (jakirkham) [inline dask/array/routines.py]: Can be optimized out if both `to_begin` and `to_end` are `None` using the proposed change to `concatenate` in PR ( https://github.com/dask/dask/pull/2610 ).

**Comment by jakirkham** [comment]
Any thoughts on this one?

**Comment by jcrist** [comment]
Looks good to me. Thanks @jakirkham!

**Comment by jakirkham** [comment]
Thanks @jcrist.

**Pull request merged**